enhancements/authentication/improved-login-workflow.md

```diff
@@ -0,0 +1,319 @@
+and [PKCE](https://datatracker.ietf.org/doc/html/rfc7636).
+
+## Motivation
+
+Currently, when logging in with the `oc` client a user has to perform the following steps:
+
+1. Open the console in the browser.
+2. Login to console with the configured identity provider.
+3. Click on the **"Copy Login Command"** button which redirects the page to the OAuth application.
+4. Perform login again like in step 2.
+5. Copy the displayed login command.
+6. Run the copied command to configure the client to configure the client and access the cluster.
+
+This process can be simplified by using OAuth Authorization Grants for configuring `oc` without having to perform a
+login multiple times and also without copy-pasting the login command with the token.
+
+### Goals
+
+Provide a single-step login in `oc` for identity providers that don't post HTTP challenges.
+
+### Non-Goals
+
+TBD
+
+## Proposal
+
+### User Stories
+
+TBD
+
+### API Extensions
+
+None
+
+### Implementation Details [optional]
+
+#### _oc login_ command
+
+The oc login command will have a new flag `--auto` which indicates to `oc` that it should use the Oauth2 Authorization
+Flow to login.
+
+#### OAuth server
+
+The internal OAuth server uses [OSIN](https://github.com/openshift/osin) as a dependency to provide OAuth2 Authorized
+Grant Flow for a registered client to fetch tokens. It also has support for PKCE(Proof Key for Code Exchange) as
+specified in [rfc7636](https://datatracker.ietf.org/doc/html/rfc7636). _rfc7636_ is an enhancement for public clients,
+i.e. clients which cannot hold client credentials to fetch an access token securely. Similarly `oc` is also a public
+client because it cannot be distributed with client credentials to retrieve the access token and PKCE is designed for
+use cases like this one.
+
+Another change which may be required is to allow the port on the callback URL to be variable when the callback is to
+localhost. This is required because the webserver might not always be able to use the same port to listen on for the
+callback webserver.
+
+#### cluster-authentication-operator
+
+The [cluster-authentication-operator](https://github.com/openshift/cluster-authentication-operator) provisions the
+default [OAuthClient](https://docs.openshift.com/container-platform/4.9/rest_api/oauth_apis/oauthclient-oauth-openshift-io-v1.html)
+resources. It already provisions a client for `oc` but sets the `respondWithChallenges` to _true_ which is unsuitable
+for OAuth Grant Flow. Therefore, another client is required where the `respondWithChallenges` is set to `false` which
+should also be provisioned by the operator.
+
+### Risks and Mitigations
+
+## Design Details
+
+### Open Questions [optional]
+
+This is where to call out areas of the design that require closure before deciding to implement the design. For
+instance,
+> 1. This requires exposing previously private resources which contain sensitive information. Can we do this?
```

Comment:
That looks like a security risk to me.

Reply:
I've added how callback to localhost can be a risk and how it's mitigated with PKCE.